### ISSUE TITLE
Documentation Request for deferring Close()

### ISSUE THREAD

**Issue by Zamiell** [ryanrolds/sqlclosecheck#8]
Hello, and thanks for the linter.

sqlclosecheck complains that `rows.Close()` should use defer. Can you explain **why** this is? For example, I don't see a difference between:

```
rows, _ := db.Query("SELECT username FROM users")
for rows.Next() {
	// Do something
}
rows.Close()
return
```

and

```
rows, _ := db.Query("SELECT username FROM users")
for rows.Next() {
	// Do something
}
defer rows.Close()
return
```

Putting defer here just seems confusing / a mistake / superfluous.

**Comment by ryanrolds** [comment]
Using a defer is preferred because if someone comes in later and puts an early return and doesn't make sure to also call Close then connection exhaustion is possible. This is a severe enough problem that is easy to miss that enforcing a defer is IMO a good course of action.

I agree that this linter should make sure that the defer is called immediately or very closely after a function returns the results/tx. I also agree that requiring a defer is stylistic. I will consider removing the defer requirement and instead check all branches call close.

**Comment by Zamiell** [comment]
> Using a defer is preferred because if someone comes in later and puts an early return and doesn't make sure to also call Close then connection exhaustion is possible.

If that is the case, then I propose that the linter should **also** require that the "defer rows,Close()" is put immediately on the line after the "rows := ..." line.
(As opposed to just complaining that you have a "rows.Close()" somewhere.)

**Comment by ryanrolds** [comment]
I 100% agree and I've added these improvements to my list. The next step for me is to research how to enforce these rules as I'm fairly new to the `ssa` package.

**Comment by ammanley** [comment]
Not to crash your thread, but can I ask for a quick clarification? 

Are we saying that we want to:
- enforce that a `defer rows.Close()` is happening _immediately_ after the `rows, err := db.Query('someSQLhere')`?

And/Or:
- check all branches in the func to enforce that `defer rows.Close()` is happening _somewhere_, but just always before any (early) returns?

Would something along the lines of:
```
var rows *sql.Rows
defer rows.Close()

rows, err := db.Query('someSQLhere')
...function carries on
```
be suitable? Given we are defering the call to close before the chance for any early returns is introduced.

**Comment by ryanrolds** [comment]
It would not be immediately after `db.Query(...)` as the error should be checked first. If `err` *isn't* `nil` then `rows` should be `nil`. The plan is to have the defer right after the error check (when an error is returned with the cursor).

My fear with the last example is that if the error check returns and doesn't do something to make `rows` not `nil` then that logic may panic. But, maybe my in-head Go interpreter isn't right.

Glad I could help. :)

**Comment by freak12techno** [comment]
I have a kind of a false positive there:
```
for _, query := range queries {
	statement, err := db.Prepare(query)
	if _, err := statement.Exec(); err != nil {
		panic(err)
	}

	statement.Close()
}
```

If using defer, it won't work correctly as it's executed at the end of the function, as Goland suggests, and if not using it, the linter would complain.

**Comment by ryanrolds** [comment]
Defer in a loop can be tricky. A solution:
```go
for _, query := range queries {
	func() {   
		statement, err := db.Prepare(query)
		if err != nil {
			panic(err)
		}
		defer statement.Close()

		if _, err := statement.Exec(); err != nil {
			panic(err)
		}
	}()
}
```
